053-configurable-event-store/proposal.md

```diff
@@ -0,0 +1,457 @@
+        notifier,     // but subscribe to the Postgres notification bus to know when new events arrive
+        from,
+    ), nil
+}
+```
+
+I've toyed with the idea of having an optional interface that `EventStore`
+implementations can implement that would overwrite this default...something
+like:
+
+```go
+type EventNotificationBus interface {
+    Notify(build Build) error
+    Listen(build Build) (<-chan struct{}, error)
+}
+```
+
+But I'm not sure if there's much value in that.
+
+
+## <a name="migrations">Migrations</a>
+
+Currently, we have migrations to set up the `build_events` table, as well as
+`TRIGGER`s for creating/deleting `pipeline_build_events_x` and
+`team_build_events_x` tables when pipelines and teams are created/deleted.
+
+I propose that we comment out these up migrations (so future deployments don't
+get these tables/triggers), and leave any database initialization (or whatever
+appropriate backend setup) to `BuildEvent.Setup()`. This means `Setup` will
+have to be "smart" (i.e. have to handle the cases where the initialization has
+already been applied). It also makes it difficult to make changes to the
+schema, unlike with migrations - each `EventStore` (that stores data in a
+structured way - this may just end up being Postgres) will need to basically
+implement its own form of migrations if we need to change the schema. My guess
+is that we won't be changing the schema too often, though (given that we
+haven't changed `build_events` since the initial migration).
+
+
+## Configuration
+
+One thing I haven't really explored is how operators would configure the
+backend store. I think it could be similar to how Credential managers are
+configured - there are groups of independent configuration options for each
+possible `EventStore`. If certain flags are configured for a given backend, use
+that backend.
+
+
+# Open Questions
+
+* Does `Key` need to be `interface{}`, or can it be more specific like `uint`?
+  For backends that don't support auto-incrementing, but still need some
+  ordering (if the timestamp isn't sufficient), we **could** create a Postgres
+  sequence for every backend (external to the `EventStore`), and change the
+  signature to `Put(build db.Build, eventID uint, event atc.Event)`
+
+* Do we care about the `Last-Event-ID` header on the
+  `/api/v1/builds/{build_id}/events` endpoint? I don't think we use it anywhere
+  internally, but I guess we may want to keep it for backwards compatibility?
+  It's not exposed by `fly` or `go-concourse`, either.
+
+* Would any event stores benefit from bringing their own notification bus
+  implementation, rather than relying on Postgres' notification bus? Does using
+  the Postgres notification bus defeat the purpose of having an external store in
+  any way (e.g. could this be a bottleneck)?
+
+* Users won't be able to access build events that exist in the main Postgres
+  database if they switch to a new backend store. How can we make the
```

Comment:
I think there are two use cases being mixed here.

The first is "tell me what just broke", the thing Concourse UX is optimised for. That data works best in PostgreSQL, close to the ATC.

The second is "I need a complete archive of everything ever". Concourse enables that but isn't really putting it front and centre. The overhead has been historically bad enough that the archival need was at cross-purposes with the diagnostic goal. This is where draining to secondary stores is useful. 

If you have a secondary store _in addition_ to the main store, then setting short build histories on the _main_ store, but not the _secondary_ store, is a relatively safe compromise for most installations.

Reply:
Yeah, that assessment makes sense. I think the proposed `EventStore` interface should be able to support this (similar to the "Cold Storage" example I provided, but could drain to an arbitrary secondary `EventStore`). Pair that with build log reaping to clear out old Postgres logs, and we can keep recent builds close to the ATC in Postgres while archiving everything in a secondary store (which could be S3, Elasticsearch, etc)